8.0 RC 1 regression:  tartree does not work for .tar.gz files

Tartree seems to be broken in 8.0-RC1

The following code works in gradle 7.6 but throws an error in 8.0-RC1. 

```
plugins {
    id "de.undercouch.download" version "5.3.0"
}

tasks.register("download", Download) {
    src("https://github.com/sass/dart-sass/releases/download/1.57.1/dart-sass-1.57.1-linux-x64.tar.gz")
    dest file("$buildDir")
}

tasks.register("unpack", Copy) {
    dependsOn tasks.named("download")
    from tarTree("$buildDir/dart-sass-1.57.1-linux-x64.tar.gz")
    into file("$buildDir/somefolder")
}
```
`./gradlew unpack`



### Expected Behavior
File should be unpacked

```
### Current Behavior

> Task :download
Download https://github.com/sass/dart-sass/releases/download/1.57.1/dart-sass-1.57.1-linux-x64.tar.gz
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.

> Task :unpack FAILED
2 actionable tasks: 2 executed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':unpack'.
> Unable to expand TAR 'C:\proyectos\gradle-jartree\build\dart-sass-1.57.1-linux-x64.tar.gz'
    The tar might be corrupted or it is compressed in an unexpected way.
    By default the tar tree tries to guess the compression based on the file extension.
    If you need to specify the compression explicitly please refer to the DSL reference.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

```


### Context


### Steps to Reproduce 
I provided is a minimal reproducer

### Your Environment
Tested in windows and linux, in java 17 and 18
Build scan URL:
https://gradle.com/s/gipau4zcwhi3e

--- a/subprojects/core/src/main/java/org/gradle/api/internal/file/archive/TarFileTree.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/file/archive/TarFileTree.java
@@ -163,11 +163,12 @@ public class TarFileTree extends AbstractArchiveFileTree {
 
         if (signatureLength >= tarHeaderSize) {
             try (TarArchiveInputStream tais = new TarArchiveInputStream(new ByteArrayInputStream(tarHeader))) {
-                if (tais.getNextTarEntry() == null) {
+                TarArchiveEntry tarEntry = tais.getNextTarEntry();
+                if (tarEntry == null) {
                     // empty TAR
                     return;
                 }
-                if (tais.getNextTarEntry().isCheckSumOK()) {
+                if (tarEntry.isCheckSumOK()) {
                     return;
                 }
             } catch (Exception e) {
@@ -230,7 +231,7 @@ public class TarFileTree extends AbstractArchiveFileTree {
             }
 
             @Override
-            public void close() throws IOException {
+            public void close() {
             }
         }
     }

--- a/subprojects/core/src/test/groovy/org/gradle/api/internal/file/archive/TarFileTreeTest.java
+++ b/subprojects/core/src/test/groovy/org/gradle/api/internal/file/archive/TarFileTreeTest.java
@@ -27,12 +27,7 @@ import org.gradle.cache.internal.TestCaches;
 import org.gradle.test.fixtures.file.TestFile;
 import org.gradle.util.TestUtil;
 import org.junit.Test;
-import spock.lang.Issue;
 
-import java.io.InputStream;
-import java.net.URL;
-import java.nio.file.Files;
-import java.nio.file.StandardCopyOption;
 import java.util.HashMap;
 import java.util.Map;
 
@@ -111,20 +106,6 @@ public class TarFileTreeTest extends AbstractArchiveFileTreeTest {
 
         assertVisits(tree, toList("subdir/file1.txt", "subdir2/file2.txt"), toList("subdir", "subdir2"));
         assertSetContainsForAllTypes(tree, toList("subdir/file1.txt", "subdir2/file2.txt"));
-    }
-
-    @Test
-    @Issue("https://github.com/gradle/gradle/issues/23391")
-    public void readsTruncatedGzippedTarFile() throws Exception {
-        TestFile tgz = tempDirProvider.getTestDirectory().file("test.tgz");
-
-        InputStream in = new URL("https://github.com/sass/dart-sass/releases/download/1.57.1/dart-sass-1.57.1-linux-x64.tar.gz").openStream();
-        Files.copy(in, tgz.toPath(), StandardCopyOption.REPLACE_EXISTING);
-
-        MaybeCompressedFileResource resource = new MaybeCompressedFileResource(new LocalResourceAdapter(TestFiles.fileRepository().localResource(tgz)));
-        TarFileTree tree = new TarFileTree(asProvider(tgz), asProvider(resource), fileSystem(), directoryFileTreeFactory(), fileHasher(), TestCaches.decompressionCache(tempDirProvider.getTestDirectory().createDir("cache-dir")));
-
-        assertSetContainsForAllTypes(tree, toList("dart-sass/sass", "dart-sass/src/LICENSE"));
     }
 
     @Test

--- a/subprojects/core/src/test/groovy/org/gradle/api/internal/file/archive/TarFileTreeTest.java
+++ b/subprojects/core/src/test/groovy/org/gradle/api/internal/file/archive/TarFileTreeTest.java
@@ -27,7 +27,12 @@ import org.gradle.cache.internal.TestCaches;
 import org.gradle.test.fixtures.file.TestFile;
 import org.gradle.util.TestUtil;
 import org.junit.Test;
+import spock.lang.Issue;
 
+import java.io.InputStream;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
 import java.util.HashMap;
 import java.util.Map;
 
@@ -109,6 +114,20 @@ public class TarFileTreeTest extends AbstractArchiveFileTreeTest {
     }
 
     @Test
+    @Issue("https://github.com/gradle/gradle/issues/23391")
+    public void readsTruncatedGzippedTarFile() throws Exception {
+        TestFile tgz = tempDirProvider.getTestDirectory().file("test.tgz");
+
+        InputStream in = new URL("https://github.com/sass/dart-sass/releases/download/1.57.1/dart-sass-1.57.1-linux-x64.tar.gz").openStream();
+        Files.copy(in, tgz.toPath(), StandardCopyOption.REPLACE_EXISTING);
+
+        MaybeCompressedFileResource resource = new MaybeCompressedFileResource(new LocalResourceAdapter(TestFiles.fileRepository().localResource(tgz)));
+        TarFileTree tree = new TarFileTree(asProvider(tgz), asProvider(resource), fileSystem(), directoryFileTreeFactory(), fileHasher(), TestCaches.decompressionCache(tempDirProvider.getTestDirectory().createDir("cache-dir")));
+
+        assertSetContainsForAllTypes(tree, toList("dart-sass/sass", "dart-sass/src/LICENSE"));
+    }
+
+    @Test
     public void failsWhenArchiveFileDoesNotExist() {
         try {
             tree.visit(new EmptyFileVisitor());
@@ -160,7 +179,7 @@ public class TarFileTreeTest extends AbstractArchiveFileTreeTest {
     public void readsTarFileWithNullPermissions() {
         resources.findResource("nullpermissions.tar").copyTo(archiveFile);
 
-        final Map<String, Integer> expected = new HashMap<String, Integer>();
+        final Map<String, Integer> expected = new HashMap<>();
         expected.put("bin", 0755);
 
         assertVisitsPermissions(tree, expected);

--- a/subprojects/soak/src/integTest/groovy/org/gradle/file/archive/TarFileTreeSoakTest.groovy
+++ b/subprojects/soak/src/integTest/groovy/org/gradle/file/archive/TarFileTreeSoakTest.groovy
@@ -24,12 +24,10 @@ import org.gradle.api.internal.file.archive.TarFileTree
 import org.gradle.api.provider.Provider
 import org.gradle.api.resources.internal.LocalResourceAdapter
 import org.gradle.cache.internal.TestCaches
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
 import org.gradle.test.fixtures.file.TestFile
-import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
 import org.gradle.util.TestUtil
-import org.junit.Rule
 import spock.lang.Issue
-import spock.lang.Specification
 
 import java.nio.file.Files
 import java.nio.file.StandardCopyOption
@@ -39,10 +37,7 @@ import static org.gradle.api.internal.file.TestFiles.fileHasher
 import static org.gradle.api.internal.file.TestFiles.fileSystem
 import static org.gradle.util.internal.WrapUtil.toList
 
-class TarFileTreeSoakTest extends Specification {
-
-    @Rule
-    public final TestNameTestDirectoryProvider temporaryFolder = new TestNameTestDirectoryProvider(getClass())
+class TarFileTreeSoakTest extends AbstractIntegrationSpec {
 
     @Issue("https://github.com/gradle/gradle/issues/23391")
     def "can decompress TAR archive with truncated header"() {

--- a/subprojects/soak/src/integTest/groovy/org/gradle/file/archive/TarFileTreeSoakTest.groovy
+++ b/subprojects/soak/src/integTest/groovy/org/gradle/file/archive/TarFileTreeSoakTest.groovy
@@ -49,8 +49,9 @@ class TarFileTreeSoakTest extends Specification {
         given:
         TestFile tgz = temporaryFolder.file("test.tgz")
 
-        InputStream input = new URL("https://github.com/sass/dart-sass/releases/download/1.57.1/dart-sass-1.57.1-linux-x64.tar.gz").openStream()
-        Files.copy(input, tgz.toPath(), StandardCopyOption.REPLACE_EXISTING)
+        try (InputStream input = new URL("https://github.com/sass/dart-sass/releases/download/1.57.1/dart-sass-1.57.1-linux-x64.tar.gz").openStream()) {
+            Files.copy(input, tgz.toPath(), StandardCopyOption.REPLACE_EXISTING)
+        }
 
         when:
         MaybeCompressedFileResource resource = new MaybeCompressedFileResource(new LocalResourceAdapter(TestFiles.fileRepository().localResource(tgz)))

--- a/subprojects/soak/src/integTest/groovy/org/gradle/file/archive/TarFileTreeSoakTest.groovy
+++ b/subprojects/soak/src/integTest/groovy/org/gradle/file/archive/TarFileTreeSoakTest.groovy
@@ -1,0 +1,78 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.file.archive
+
+import org.gradle.api.file.EmptyFileVisitor
+import org.gradle.api.file.FileVisitDetails
+import org.gradle.api.internal.file.MaybeCompressedFileResource
+import org.gradle.api.internal.file.TestFiles
+import org.gradle.api.internal.file.archive.TarFileTree
+import org.gradle.api.provider.Provider
+import org.gradle.api.resources.internal.LocalResourceAdapter
+import org.gradle.cache.internal.TestCaches
+import org.gradle.test.fixtures.file.TestFile
+import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
+import org.gradle.util.TestUtil
+import org.junit.Rule
+import spock.lang.Issue
+import spock.lang.Specification
+
+import java.nio.file.Files
+import java.nio.file.StandardCopyOption
+
+import static org.gradle.api.internal.file.TestFiles.directoryFileTreeFactory
+import static org.gradle.api.internal.file.TestFiles.fileHasher
+import static org.gradle.api.internal.file.TestFiles.fileSystem
+import static org.gradle.util.internal.WrapUtil.toList
+
+class TarFileTreeSoakTest extends Specification {
+
+    @Rule
+    public final TestNameTestDirectoryProvider temporaryFolder = new TestNameTestDirectoryProvider(getClass())
+
+    @Issue("https://github.com/gradle/gradle/issues/23391")
+    def "can decompress TAR archive with truncated header"() {
+        given:
+        TestFile tgz = temporaryFolder.file("test.tgz")
+
+        InputStream input = new URL("https://github.com/sass/dart-sass/releases/download/1.57.1/dart-sass-1.57.1-linux-x64.tar.gz").openStream()
+        Files.copy(input, tgz.toPath(), StandardCopyOption.REPLACE_EXISTING)
+
+        when:
+        MaybeCompressedFileResource resource = new MaybeCompressedFileResource(new LocalResourceAdapter(TestFiles.fileRepository().localResource(tgz)))
+        TarFileTree tree = new TarFileTree(asProvider(tgz), asProvider(resource), fileSystem(), directoryFileTreeFactory(), fileHasher(), TestCaches.decompressionCache(temporaryFolder.createDir("cache-dir")))
+        def files = getFiles(tree)
+
+        then:
+        files.containsAll(toList("sass", "LICENSE"))
+    }
+
+    private static Set<String> getFiles(TarFileTree fileTree) {
+        final Set<String> files = new LinkedHashSet<>()
+        fileTree.visit(new EmptyFileVisitor() {
+            @Override
+            void visitFile(FileVisitDetails fileDetails) {
+                files.add(fileDetails.getName())
+            }
+        })
+        return files
+    }
+
+    private static <T> Provider<T> asProvider(T object) {
+        return TestUtil.providerFactory().provider(() -> object)
+    }
+}

